src/NUnitFramework/framework/Internal/ValueGenerator.cs

```diff
@@ -0,0 +1,253 @@
+// ***********************************************************************
+// Copyright (c) 2018 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using NUnit.Compatibility;
+
+namespace NUnit.Framework.Internal
+{
+    internal abstract class ValueGenerator<T> : ValueGenerator
+    {
+        private static Exception CreateNotSupportedException(string description)
+        {
+            return new NotSupportedException($"{typeof(T)} is using the default value generator which does not support {description}.");
+        }
+
+        public virtual int Compare(T x, T y)
```

Comment:
Do you already have an idea for which types we would need to override this? (or is it just to make it possible to extend this class later on).

Reply:
Also, this is a correct but less efficient default implementation. Each comparer could override this with `=> x.CompareTo(y);` for a performance boost since neither `!typeof(IComparable<T>).IsAssignableFrom(typeof(T))` nor `Comparer<T>.Default.Compare(x, y)` are detected as constant by the JIT. Because I would rather keep it simple and wait to add code till a performance need is identified, I didn't override it. So this is kind of a note for the future.

As to whether I should have left the method sealed, maybe? But if we ever create a value generator for things that aren't IComparable, it'll need to be overridden. I don't know, do you have a feeling one way or the other?

Reply:
This is a correct but less efficient default implementation. Each comparer could override this with `=> x.CompareTo(y);` for a performance boost since neither `!typeof(IComparable<T>).IsAssignableFrom(typeof(T))` nor `Comparer<T>.Default.Compare(x, y)` are detected as constant by the JIT. Because I would rather keep it simple and wait to add code till a performance need is identified, I didn't override it. So this is kind of a note for the future.

As to whether I should have left the method sealed until such need arises, maybe? If we ever create a value generator for things that aren't IComparable, it'll need to be overridden. I don't know, do you have a feeling one way or the other?

Reply:
I think it is fine as it is. I was just wondering when I saw `virtual`.

Reply:
Okay, thanks for commenting. Sounds like I'm good to merge then. 👍